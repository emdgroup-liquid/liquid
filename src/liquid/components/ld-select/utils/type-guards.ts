export const isLdOption = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptionElement | HTMLLdOptionInternalElement =>
  ["LD-OPTION", "LD-OPTION-INTERNAL"].includes((el as HTMLElement)?.tagName);

export const isLdOptgroup = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptgroupElement | HTMLLdOptgroupInternalElement =>
  ["LD-OPTGROUP", "LD-OPTGROUP-INTERNAL"].includes(
    (el as HTMLElement)?.tagName,
  );

export const isLdOptionInternal = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptionInternalElement =>
  ["LD-OPTION-INTERNAL"].includes((el as HTMLElement)?.tagName);

export const isLdOptgroupInternal = (
  el: HTMLElement | Node | EventTarget,
): el is HTMLLdOptgroupInternalElement =>
  ["LD-OPTGROUP-INTERNAL"].includes((el as HTMLElement)?.tagName);

type HTMLLdOptInternal =
  | HTMLLdOptionInternalElement
  | HTMLLdOptgroupInternalElement;
export const isLdOptInternalHidden = (
  opt: HTMLLdOptInternal,
): opt is
  | (HTMLLdOptInternal & {
      hidden: true;
    })
  | (HTMLLdOptInternal & {
      filtered: true;
    }) => {
  return opt.hidden || opt.filtered;
};
