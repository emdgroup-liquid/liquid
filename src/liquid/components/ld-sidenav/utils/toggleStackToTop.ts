import { LdSidenavNavitem } from '../ld-sidenav-navitem/ld-sidenav-navitem'

export const toggleStackToTop = (el: HTMLElement, stacked: boolean) => {
  const outerSlider = el.closest('ld-sidenav-slider')
  if (!outerSlider) return

  const sidenavPaddingY =
    parseInt(
      window
        .getComputedStyle(outerSlider)
        .getPropertyValue('--ld-sidenav-padding-y')
    ) * 16

  // If not stacked, put everything back in place.
  if (!stacked) {
    el.style.removeProperty('--ld-slider-navitem-move-up')
    return
  }

  // Else, calculate how much vertical space is taken by stacked items above
  // and move the nav item up accordingly.
  let totalSpaceAbove = 0
  let totalSpaceOccupiedAbove = 0
  for (const elem of Array.from(el.parentElement.children)) {
    if (el === elem) break

    if (
      elem.tagName === 'LD-SIDENAV-SEPARATOR' ||
      (elem.tagName === 'LD-SIDENAV-NAVITEM' &&
        !['secondary', 'tertiary'].includes(
          (elem as unknown as LdSidenavNavitem).mode
        ))
    ) {
      totalSpaceOccupiedAbove += elem.getBoundingClientRect().height
    } else {
      totalSpaceAbove += sidenavPaddingY
    }
    totalSpaceAbove += elem.getBoundingClientRect().height
  }

  const spaceToMoveUp = totalSpaceAbove - totalSpaceOccupiedAbove
  el.style.setProperty('--ld-slider-navitem-move-up', `-${spaceToMoveUp}px`)
}
