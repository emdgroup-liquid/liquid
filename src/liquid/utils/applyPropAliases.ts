/**
 * Allows using lowercase props as aliases for camelcase props.
 * This resolves issues in react based applications, where
 * otherwise only kebabcase attributes would work, but are not
 * proposed by intellisense.
 */
export function applyPropAliases() {
  if (!this.el) {
    throw new Error(`Component ${this.name} is missing @Element() decorator.`)
  }
  for (const prop in this) {
    if (prop !== prop.toLowerCase()) {
      const lowercaseAttr = this.el.getAttribute(prop.toLowerCase())
      if (lowercaseAttr) {
        this[prop] = lowercaseAttr
      }
    }
  }
}
