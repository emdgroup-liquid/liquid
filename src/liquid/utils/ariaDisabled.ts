// 'true', '1', 'yolo' etc. correspond to an aria disabled element, but not 'false'.
export const isAriaDisabled = (ariaDisabledValue: string | undefined) =>
  ariaDisabledValue && ariaDisabledValue !== "false";
