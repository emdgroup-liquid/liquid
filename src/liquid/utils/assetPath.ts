import { Build } from '@stencil/core'
import { version } from '../../../package.json'

let missingAssetPathWarningDisplayed = false

/**
 * Reads components asset path config from meta tag or global variable.
 * This is a workaround until the following issues are resolved:
 * https://github.com/ionic-team/stencil/issues/3470
 * https://github.com/ionic-team/stencil-ds-output-targets/issues/186
 */
export const getAssetPath = (path: string) => {
  // Get asset path from meta tag if available
  const metaLdAssetPath = document.head.querySelector<HTMLMetaElement>(
    'meta[data-ld-asset-path]'
  )?.dataset.ldAssetPath
  // Get asset path from window if available
  const windowLdAssetPath = window.__LD_ASSET_PATH__
  // Uses CDN as fallback if no asset path is set
  const cdnAssetPath = `https://cdn.jsdelivr.net/npm/@emdgroup-liquid/liquid${
    version ? '@' + version : ''
  }/dist/liquid/`

  const assetBasePath = Build.isTesting
    ? '/dist/liquid'
    : metaLdAssetPath || windowLdAssetPath || cdnAssetPath || '/'

  // Display warning if assets are fetched from CDN. This is only displayed once.
  if (
    assetBasePath.startsWith('https://cdn.jsdelivr.net/npm/') &&
    !missingAssetPathWarningDisplayed
  ) {
    missingAssetPathWarningDisplayed = true
    console.warn(
      `Fetching Liquid Oxygen assets from jsDelivr CDN.\n\nWe recommend bundling Liquid Oxygen assets with your application and setting the asset path accordingly.\n\nFor more information see the documentation:\nhttps://liquid.merck.design/liquid/guides/component-assets/`
    )
  }

  let assetPath = path

  if (path.startsWith('./')) {
    assetPath = path.substring(2)
  }

  if (!assetBasePath.endsWith('/')) {
    assetPath = '/' + assetPath
  }

  return assetBasePath + assetPath
}
