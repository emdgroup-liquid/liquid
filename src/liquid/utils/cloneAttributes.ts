function isFalsy(value: string | null | undefined) {
  return value === undefined || value === null || value === 'false'
}

export function cloneAttributes(attributesToIgnore: string[] = []) {
  const attributesToIgnoreSet = new Set([
    'style',
    'id',
    'class',
    'slot',
    'part',
    ...attributesToIgnore,
  ])

  // Get attributes not in props.
  const attributesToClone = {}
  for (const attr of this.el.attributes) {
    if (attributesToIgnoreSet.has(attr.name) || isFalsy(attr.value)) {
      continue
    }
    const valueToClone = attr.value === '' ? true : attr.value
    attributesToClone[attr.name] = valueToClone
  }

  // Update cloned attributes state.
  this.clonedAttributes = attributesToClone

  // Set up attributes observer.
  const callback = (mutationsList) => {
    // Update cloned attributes state.
    for (const mutation of mutationsList) {
      const { attributeName } = mutation
      if (!attributesToIgnoreSet.has(attributeName)) {
        const attrValue = this.el.getAttribute(attributeName)
        if (isFalsy(attrValue)) {
          delete this.clonedAttributes[attributeName]
        } else {
          this.clonedAttributes[attributeName] = attrValue
        }
      }
    }

    // Trigger state update.
    this.clonedAttributes = { ...this.clonedAttributes }
  }
  const observer = new MutationObserver(callback)
  observer.observe(this.el, {
    subtree: false,
    childList: false,
    attributes: true,
  })

  return observer
}
