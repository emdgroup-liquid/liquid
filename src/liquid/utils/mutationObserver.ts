import MutationObserver from "mutation-observer";

let triggerableMutationObservers = [];
const TriggerableMutationObserver = function (cb) {
  const triggerableMutationObserver = new MutationObserver(cb);
  triggerableMutationObserver.trigger = cb;
  triggerableMutationObservers.push(triggerableMutationObserver);
  return triggerableMutationObserver;
};

global.MutationObserver = TriggerableMutationObserver as MutationObserver;

export function getTriggerableMutationObservers() {
  return triggerableMutationObservers;
}

export function clearTriggerableMutationObservers() {
  triggerableMutationObservers = [];
}
