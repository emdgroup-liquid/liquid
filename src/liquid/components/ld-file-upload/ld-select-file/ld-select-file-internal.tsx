import {
  Component,
  h,
  Host,
  Prop,
  Element,
  State,
  Event,
  EventEmitter,
  Fragment,
} from '@stencil/core'
import { getClassNames } from '../../../utils/getClassNames'
import { getAssetPath } from '../../../utils/assetPath'
import type { LdUploadItem } from '../ld-file-upload'

let selectFileCount = 0

/**
 * @internal
 * @virtualProp ref - reference to component
 */
@Component({
  tag: 'ld-select-file-internal',
  styleUrl: 'ld-select-file-internal.shadow.css',
  shadow: true,
  assetsDirs: ['assets'],
})
export class LdSelectFileInternal {
  @Element() el: HTMLLdSelectFileInternalElement
  private idPrefix = `ld-select-file-${++selectFileCount}`

  /** Max. file size in bytes */
  @Prop() maxFileSize?: number

  /** Defines whether upload starts immediately after selecting files or after confirmation. */
  @Prop() immediate?: boolean = false

  /** Defines whether start upload button has been clicked (only relevant when immediate is false)
   * @internal
   */
  @Prop() startUploadClicked?: boolean = false

  /** Defines whether selection of multiple input files is allowed. */
  @Prop() multiple?: boolean = false

  /** Defines whether the total progress of all uploading files will be shown in the progress button */
  @Prop() showProgress?: boolean = false

  /** Selected files from the parent component */
  @Prop() uploadItems: LdUploadItem[] = []

  /** Label to be used as a header with instructions for drag and drop or file upload. */
  @Prop() labelDragInstructions = `Drag your file${
    this.multiple ? '(s)' : ''
  } here or browse`

  /** Label to be used to describe upload constraints like the maximum file size. */
  @Prop() labelUploadConstraints = `${
    this.maxFileSize !== undefined ? 'max. $maxFileSize file size' : ''
  }`

  /** Label to be used for the select files button. */
  @Prop() labelSelectFile = `Select ${this.multiple ? '' : 'a'} file${
    this.multiple ? '(s)' : ''
  }`

  /** Label to be used for the upload files button. */
  @Prop() labelUploadFile = `Upload ${this.multiple ? '' : 'a'} file${
    this.multiple ? '(s)' : ''
  }`

  /** Label to be used for the upload state header. */
  @Prop() labelUploadState = `Upload state:`

  /** Label to be used to count the amount of files that have been uploaded. */
  @Prop() labelUploadCount = `$filesUploaded of $filesTotal file${
    this.multiple ? 's' : ''
  } uploaded.`

  /** Label to be used to show the total upload percentage. */
  @Prop() labelUploadPercentage = `$uploadProgress % uploaded.`

  /** Represents whether a file is currently being dragged over the drop area */
  @State() highlighted? = false

  /** @internal */
  /** Emitted after dropping a file in the drop area. */
  @Event() ldselectfiles: EventEmitter<FileList>

  /** @internal */
  /** Emitted on upload click. */
  @Event() lduploadclick: EventEmitter

  private bytesToSize = (bytes: number) => {
    const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB']

    let sizeIndex = 0

    while (bytes >= 1024 && sizeIndex < sizes.length - 1) {
      bytes /= 1024
      sizeIndex++
    }

    const roundedSize = bytes.toFixed(2)

    return `${roundedSize} ${sizes[sizeIndex]}`
  }

  private handleUploadClick = () => {
    this.lduploadclick.emit()
  }

  private handleDragEnter = (ev: DragEvent) => {
    ev.preventDefault()
    const fileList = ev.dataTransfer.files
    if (
      (this.multiple || fileList.length <= 1) &&
      (this.immediate || !this.startUploadClicked)
    ) {
      this.highlighted = true
    }
  }

  private handleDragOver = (ev: DragEvent) => {
    ev.preventDefault()
    const fileList = ev.dataTransfer.files
    if (
      (this.multiple || fileList.length <= 1) &&
      (this.immediate || !this.startUploadClicked)
    ) {
      this.highlighted = true
    }
  }

  private handleDragLeave = (ev: DragEvent) => {
    ev.preventDefault()
    // ev.stopPropagation()
    this.highlighted = false
  }

  private handleDrop = (ev: DragEvent) => {
    ev.preventDefault()
    // ev.stopPropagation()
    this.highlighted = false

    const fileList = ev.dataTransfer.files

    if (
      (this.multiple || fileList.length <= 1) &&
      (this.immediate || !this.startUploadClicked)
    ) {
      this.ldselectfiles.emit(fileList)
    }
  }

  render() {
    const cl = getClassNames([
      'ld-select-file',
      this.highlighted && 'ld-select-file--highlighted',
    ])

    // TODO: the exact function already exists in ld-file-upload.tsx
    const calculateTotalProgress = () => {
      const activeUploads = this.uploadItems.filter(
        (item) =>
          item.state === 'pending' ||
          item.state === 'paused' ||
          item.state === 'uploading' ||
          item.state === 'uploaded'
      )
      const totalSizeSum = activeUploads.reduce(
        (partialSum, file) => partialSum + file.fileSize,
        0
      )
      const uploadedSizeSum = activeUploads.reduce(
        (partialSum, file) =>
          partialSum + file.fileSize * (file.progress / 100),
        0
      )
      const totalProgress = uploadedSizeSum / totalSizeSum || 0
      // returns total progress between 0 and 1
      return totalProgress
    }

    return (
      <Host
        class={cl}
        onDragEnter={this.handleDragEnter}
        onDragOver={this.handleDragOver}
        onDragLeave={this.handleDragLeave}
        onDrop={this.handleDrop}
      >
        <div class="ld-select-file__content">
          <img
            class="ld-select-file__image"
            src={getAssetPath('./assets/file-upload.svg')}
            alt=""
            width="142"
            height="122"
          />

          {this.immediate || (!this.immediate && !this.startUploadClicked) ? (
            <Fragment>
              <ld-typo id={`${this.idPrefix}-instructions`} variant="h5">
                {this.labelDragInstructions}
              </ld-typo>
              {this.labelUploadConstraints !== '' && (
                <ld-typo>
                  {this.labelUploadConstraints.replace(
                    '$maxFileSize',
                    this.bytesToSize(this.maxFileSize)
                  )}
                </ld-typo>
              )}
              <slot></slot>
              <ld-button
                aria-labelledby={`${this.idPrefix}-instructions`}
                class="ld-select-file__upload-button"
                size="sm"
                onClick={this.handleUploadClick}
              >
                {this.immediate ? this.labelUploadFile : this.labelSelectFile}
              </ld-button>
            </Fragment>
          ) : (
            <Fragment>
              <ld-typo variant="h5">{this.labelUploadState}</ld-typo>
              <ld-typo>
                {this.labelUploadCount
                  .replace(
                    '$filesUploaded',
                    String(
                      this.uploadItems.filter(
                        (item) => item.state === 'uploaded'
                      ).length
                    )
                  )
                  .replace('$filesTotal', String(this.uploadItems.length))}
              </ld-typo>
              {this.showProgress ? (
                <ld-typo>
                  {this.labelUploadPercentage.replace(
                    '$uploadProgress',
                    String((calculateTotalProgress() * 100).toFixed(2))
                  )}
                </ld-typo>
              ) : undefined}
            </Fragment>
          )}
        </div>
      </Host>
    )
  }
}
