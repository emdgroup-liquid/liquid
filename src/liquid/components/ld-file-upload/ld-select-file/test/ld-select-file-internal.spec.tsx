import { newSpecPage } from '@stencil/core/testing'
import { LdSelectFileInternal } from '../ld-select-file-internal'
import '../../../../utils/mutationObserver'
import { h } from '@stencil/core'

describe('ld-select-file', () => {
  it('renders default', async () => {
    const page = await newSpecPage({
      components: [LdSelectFileInternal],
      html: `<ld-select-file-internal />`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('renders upload items', async () => {
    const page = await newSpecPage({
      components: [LdSelectFileInternal],
      template: () => (
        <ld-select-file-internal
          start-upload-clicked
          uploadItems={[
            {
              file: undefined,
              state: 'uploading',
              fileName: 'file1.png',
              fileSize: 100000,
              fileType: 'png',
              progress: 50,
            },
          ]}
        />
      ),
    })
    expect(page.root.uploadItems.length).toBe(1)
  })

  it('renders immediate', async () => {
    const page = await newSpecPage({
      components: [LdSelectFileInternal],
      html: `<ld-select-file-internal immediate />`,
    })
    expect(page.root.startUploadImmediately).toBe(true)
  })

  it('renders start-upload-clicked', async () => {
    const page = await newSpecPage({
      components: [LdSelectFileInternal],
      html: `<ld-select-file-internal start-upload-clicked />`,
    })
    expect(page.root.startUploadClicked).toBe(true)
  })

  it('renders multiple', async () => {
    const page = await newSpecPage({
      components: [LdSelectFileInternal],
      html: `<ld-select-file-internal multiple />`,
    })
    expect(page.root.multiple).toBe(true)
  })

  it('emits lduploadclick event', async () => {
    const page = await newSpecPage({
      components: [LdSelectFileInternal],
      html: `<ld-select-file-internal immediate />`,
    })
    const ldSelectFile = page.root
    const uploadButton =
      ldSelectFile.shadowRoot.querySelector<HTMLLdButtonElement>(
        'ld-button[class="ld-select-file__upload-button"]'
      )

    const lduploadclickHandler = jest.fn()
    ldSelectFile.addEventListener('lduploadclick', lduploadclickHandler)

    uploadButton.click()
    await page.waitForChanges()

    expect(lduploadclickHandler).toHaveBeenCalled()
  })

  it('adds class highlighted on drag enter', async () => {
    const page = await newSpecPage({
      components: [LdSelectFileInternal],
      html: `<ld-select-file-internal immediate multiple />`,
    })
    const ldSelectFile = page.root

    await page.waitForChanges()

    // eslint-disable-next-line compat/compat
    const ev = new DragEvent('dragenter', { bubbles: true })
    ;(ev as { dataTransfer: object }).dataTransfer = {}
    ldSelectFile.dispatchEvent(ev)

    await page.waitForChanges()

    expect(ldSelectFile).toHaveClass('ld-select-file--highlighted')
  })

  it('adds class highlighted on drag over', async () => {
    const page = await newSpecPage({
      components: [LdSelectFileInternal],
      html: `<ld-select-file-internal immediate multiple />`,
    })
    const ldSelectFile = page.root

    await page.waitForChanges()

    // eslint-disable-next-line compat/compat
    const ev = new DragEvent('dragover', { bubbles: true })
    ;(ev as { dataTransfer: object }).dataTransfer = {}
    ldSelectFile.dispatchEvent(ev)

    await page.waitForChanges()

    expect(ldSelectFile).toHaveClass('ld-select-file--highlighted')
  })

  it('removes class highlighted on drag leave', async () => {
    const page = await newSpecPage({
      components: [LdSelectFileInternal],
      html: `<ld-select-file-internal immediate multiple />`,
    })
    const ldSelectFile = page.root

    await page.waitForChanges()

    // eslint-disable-next-line compat/compat
    const ev1 = new DragEvent('dragover', { bubbles: true })
    ;(ev1 as { dataTransfer: object }).dataTransfer = {}
    ldSelectFile.dispatchEvent(ev1)

    await page.waitForChanges()

    expect(ldSelectFile).toHaveClass('ld-select-file--highlighted')

    // eslint-disable-next-line compat/compat
    const ev = new DragEvent('dragleave', { bubbles: true })
    ;(ev as { dataTransfer: object }).dataTransfer = {}
    ldSelectFile.dispatchEvent(ev)

    await page.waitForChanges()

    expect(ldSelectFile).not.toHaveClass('ld-select-file--highlighted')
  })

  it('removes class highlighted on drop and emits ldselectfiles event', async () => {
    const page = await newSpecPage({
      components: [LdSelectFileInternal],
      html: `<ld-select-file-internal immediate multiple />`,
    })
    const ldSelectFile = page.root

    const ldselectfilesHandler = jest.fn()
    ldSelectFile.addEventListener('ldselectfiles', ldselectfilesHandler)

    await page.waitForChanges()

    // eslint-disable-next-line compat/compat
    const ev1 = new DragEvent('dragover', { bubbles: true })
    ;(ev1 as { dataTransfer: object }).dataTransfer = {}
    ldSelectFile.dispatchEvent(ev1)

    await page.waitForChanges()

    expect(ldSelectFile).toHaveClass('ld-select-file--highlighted')

    // eslint-disable-next-line compat/compat
    const ev = new DragEvent('drop', { bubbles: true })
    ;(ev as { dataTransfer: object }).dataTransfer = {}
    ldSelectFile.dispatchEvent(ev)

    await page.waitForChanges()

    expect(ldSelectFile).not.toHaveClass('ld-select-file--highlighted')
    expect(ldselectfilesHandler).toHaveBeenCalled()
  })
})
