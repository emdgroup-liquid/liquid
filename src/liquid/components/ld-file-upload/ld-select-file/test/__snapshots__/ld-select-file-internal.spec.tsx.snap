// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-select-file renders default 1`] = `
<ld-select-file class="ld-select-file">
  <mock:shadow-root>
    <div class="ld-select-file--vertical ld-select-file__content">
      <img alt="" class="ld-select-file__image" height="46" src="/dist/liquid/assets/file-upload.svg" width="55">
      <div class="ld-select-file__text ld-select-file__text--vertical">
        <ld-typo id="ld-select-file-1-instructions" variant="h5">
          Drag your file here or browse
        </ld-typo>
        <slot></slot>
        <ld-button aria-labelledby="ld-select-file-1-instructions" class="ld-select-file__upload-button" size="sm">
          Select a file
        </ld-button>
      </div>
    </div>
  </mock:shadow-root>
</ld-select-file>
`;
